### PR TITLE
Poc to use safetensors

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR introduces the necessary code changes to use safetensors weight file as a primary source.

### CODE REVIEW

**Comment by LysandreJik** [comment]
Thank you for being verified!

File: src/transformers/utils/__init__.py [hunk 1]
```diff
@@ -155,6 +156,8 @@
 TF_WEIGHTS_NAME = "model.ckpt"
 FLAX_WEIGHTS_NAME = "flax_model.msgpack"
 FLAX_WEIGHTS_INDEX_NAME = "flax_model.msgpack.index.json"
+SAFE_WEIGHTS_NAME = "model.safetensors"
+SAFE_WEIGHTS_INDEX_NAME = "model.safetensors.index.json"
```
Comment (julien-c) [inline src/transformers/utils/__init__.py:160]: 🥰🥰

**Comment by Narsil** [comment]
This looks so clean !

**Review by Narsil — APPROVED** [review]
LGTM.
Right now this would be only for `torch` afaik.

We can keep it that way, and maybe later try to load it for any framework (just to reduce the amount of shared weights)

File: tests/test_modeling_common.py [hunk 2]
```diff
@@ -2980,6 +2982,24 @@ def test_legacy_load_from_url(self):
             "https://huggingface.co/hf-internal-testing/tiny-random-bert/resolve/main/pytorch_model.bin", config=config
         )
 
+    @require_safetensors
+    def test_safetensors_save_and_load(self):
+        model = BertModel.from_pretrained("hf-internal-testing/tiny-random-bert")
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            model.save_pretrained(tmp_dir, safe_serialization=True)
+            # No pytorch_mode.bin file, only a model.safetensors
```
Comment (julien-c) [inline tests/test_modeling_common.py:2990]:
```suggestion
            # No pytorch_model.bin file, only a model.safetensors
```

**Comment by julien-c** [comment]
Opened Hub Pull requests to convert the following weights:
- https://huggingface.co/roberta-base/discussions/3
- https://huggingface.co/roberta-large/discussions/1
- https://huggingface.co/gpt2/discussions/6
- https://huggingface.co/Jean-Baptiste/camembert-ner/discussions/1
- https://huggingface.co/openai/clip-vit-large-patch14/discussions/5

^will merge the 3 canonical ones to be able to test easily

**Comment by julien-c** [comment]
note that you can also test from un-merged Hub PRs, you just have to pass the `refs/pr/:id` as a `revision`, for instance for gpt2:

```python
model = AutoModelForCausalLM.from_pretrained("gpt2", revision="refs/pr/6")
```

File: src/transformers/modeling_utils.py [hunk 3]
```diff
@@ -367,6 +381,20 @@ def load_state_dict(checkpoint_file: Union[str, os.PathLike]):
...
     """
+    if checkpoint_file.endswith(".safetensors") and is_safetensors_available():
+        # Check format of the archive
+        with safe_open(checkpoint_file, framework="pt") as f:
+            metadata = f.metadata()
+        if metadata.get("format") not in ["pt", "tf", "flax"]:
+            raise OSError(
+                f"The safetensors archive passed at {checkpoint_file} does not contain the valid metadata. Make sure "
+                "you save your model with the `save_pretrained` method."
+            )
```
Comment (LysandreJik) [inline src/transformers/modeling_utils.py:392]: :+1: 

If this becomes too much of an issue, we could also take a look at providing a code snipper that would manually add the metadata to the checkpoint; but that's only if there are some complaints, the error raising looks good to me

**Review by LysandreJik — COMMENTED** [review]
Looks very good already, very clean implementation

**Pull request merged**